### PR TITLE
Feature: Add an RAII wrapper to support RPC timeouts when making blocking COM calls #458

### LINKED ISSUE #458: Feature: Add an RAII wrapper to support RPC timeouts when making blocking COM calls
See #459 for the details on this

### PR DESCRIPTION
**Pull request by dmachaj**
Closes #458

## Why is this change being made?
We have a project where we want to make blocking COM calls to another COM server process, but we want protection from the calls taking an infinite amount of time.  There is an existing, but very old, wrapper that provides this functionality.  Instead of using the old wrapper I am porting it to WIL so that we can use it in a different repo than the original and to share it more broadly because it is useful.

## Briefly summarize what changed
Add a new `wil::rpc_timeout` class to <wil/com.h> along with a few test cases for it.  This class uses a combination of `CoEnableCallCancellation`, `CoDisableCallCancellation`, and `CoCancelCall` to provide the COM functionality.  `CreateThreadpoolTimer` is used to provide a timed callback that will not block on the same thread as the blocking COM call.

This object is intended to live on the stack and have a short lifetime.  It is single-use and applies to all RPC calls made while it is alive.

## How was this change tested?
New test cases in ComTests.cpp.

### CODE REVIEW

File: include/wil/com.h
Comment (jonwis) [inline include/wil/com.h:3393]: Is the `nocancel` version the one you want? What happens if the timer fires right after the COM call returns to you but before the dtor finishes here?

**Review by jonwis — APPROVED** [review]

File: include/wil/com.h
Comment (dmachaj) [reply include/wil/com.h:3393]: The destructor would block until the callback is completed, preventing the `this` pointer from being used after destruction.  The callback doesn't do significant work so it shouldn't block for long.

File: include/wil/com.h
Comment (brialmsft) [inline include/wil/com.h:3344]: >rpc_timeout [](http://example.com/codeflow?start=6&length=11)

I've always been skeptical of this name, given that this is absolutely COM-level cancelation, and could theoretically be used in inproc scenarios which don't involve the RPC runtime at all.

File: include/wil/com.h
Comment (brialmsft) [inline include/wil/com.h:3339]: This accidentally illustrates one of the subtle gotchas of this sort of timeout: it causes the cancelation of a single topmost call, and if you don't bail out after that 1 failure and make a different hanging call (or if cancelation happens to land while you're between calls, e.g. in a loop), then this will not save you.

Presumably in this comment, remote_object is a C++/WinRT-style wrapper that will throw if BlockingCOMCall returns RPC_E_CALL_CANCELED. But if somebody naively does this same thing without the throwing wrapper, then if both calls hang, the timeout will help with the first and won't apply to the second.

File: include/wil/com.h
Comment (dmachaj) [reply include/wil/com.h:3344]: Feel free to suggest a better name.  I don't mind renaming to something more accurate.

File: include/wil/com.h
Comment (dmachaj) [reply include/wil/com.h:3339]: I realized this after sending the PR and was not able to finish fixing it up before I left for the day.  The previous internal wrapper used a periodic timer to catch additional calls if the first call did not early-return.  I am bringing that back to ensure that it covers any number of calls.  However, when I added the additional test case for that I encountered a problem that I haven't been able to address just yet.

File: include/wil/com.h
Comment (dmachaj) [reply include/wil/com.h:3344]: The problem that I've been having is that `CoCancelCall` is failing with `RPC_E_NO_CONTEXT` which seems to mean that this truly does require RPC and does not work inproc across apartments.

Comment (dmachaj) [reply include/wil/com.h:3344]: If the primary test thread is changed from MTA to STA then the RPC no context error goes away and CoCancelCall returns S_OK.  That does not seem to cause the thread to unblock, though.

Comment (brialmsft) [reply include/wil/com.h:3344]: > Feel free to suggest a better name

com_timeout?

> That does not seem to cause the thread to unblock, though.

Yeah, I was doing some testing of my own against that last night and saw the same thing. Cancelation state gets set on the call, but that doesn't result in the client thread being able to return unilaterally. If I allow the callee STA to pump messages after the cancelation, the caller does end up getting RPC_E_CALL_CANCELED without invoking the server method, but yeah it was still blocked on the server to return.

Broadly speaking, issuing a cancelation doesn't guarantee that the cancelation will be seen/respected within any specific timeframe, so letter-of-the-law this is fine. But this sort of gap definitely limits the usefulness of canceling inproc calls.

I'll also note that one can use CoSetCancelObject to allow custom cancelation even for calls that don't pass 7through the COM runtime (as another example of non-RPC cancelation).

File: include/wil/com.h
Comment (dunhor) [inline include/wil/com.h:3329]: Is there any potential issue with just including `win32_helpers.h` in this file? `com.h` is not set up to be includable more than once, making patterns like this very brittle and prone to include order issues.

Comment (dmachaj) [reply include/wil/com.h:3329]: I'm not sure.  I was hoping you and other experts could advise on that.  The filetime helpers are definitely useful in this case so I'd prefer to use them instead of doing ULARGE_INTEGER math manually.

File: include/wil/com.h
Comment (dunhor) [reply include/wil/com.h:3339]: > or if cancelation happens to land while you're between calls

The single failing call seems easy enough for consumers to handle themselves (and is arguably good COM hygiene), but this part is a little scary. I second the persistence of the periodic timer

File: include/wil/com.h
Comment (dunhor) [reply include/wil/com.h:3329]: Unless there are good, specific concerns, I'd say just include the header. If problems arise, the number of failures will determine how we might go about mitigating the issue (e.g. something like an opt-out for few failures). Not that I anticipate that, though. Risking problems is better than introducing a ticking time bomb of include order dependencies.

Comment (dmachaj) [reply include/wil/com.h:3329]: That makes sense.  I'll make that change in a future iteration.

File: include/wil/com.h
Comment (dunhor) [inline include/wil/com.h:3394]: There are several places for failure. A few things to consier:
1. Add an `operator bool` (or some sort of `HRESULT` getter) so that the caller can check if initialization succeeded.
2. Change this to a more generic `m_initResult` that will also change to the result of `GetLastError` when e.g. threadpool functions fail
3. Make non-throwing, failfast, and throwing versions of this type (advanced; not necessarily necessary)

File: include/wil/com.h
Comment (dunhor) [reply include/wil/com.h:3393]: One thing to keep in mind - although I don't necessarily think it's an issue here - is that members construct in-order before the constructor runs and destruct in reverse order after the destructor runs. Effectively, this means that the callback could be running while the rest of this object has been destructed (since this is the first member). You're not doing anything crazy like setting `m_threadId` to zero and writing to `m_timedOut` should be no-op, however this also means that it's possible for `CoDisableCallCancellation` to be called before `CoCancelCall` is called in the callback. I'd hope that COM handles this gracefully and no-ops/returns failure for `CoCancelCall` and so no harm is done; just something to consider.

Comment (dmachaj) [reply include/wil/com.h:3396]: That's a great point.  Reordering fields should eliminate that risk.  I'll do that and add a comment describing it.

File: include/wil/com.h
Comment (dmachaj) [reply include/wil/com.h:3396]: I keep going back and forth on this in my head and I haven't been able to converge on a firm opinion.  The old version that I'm porting from was "best effort" in terms of errors in the constructor.  This version continues that approach.  However, I also understand the desire to have stronger error detection if this is part of WIL.

Comment (dunhor) [reply include/wil/com.h:3396]: I'm _okay_ with it being best effort so long as it's documented as such. That said, I'm sure some consumers would want to know about failure. Perhaps something like the following would suffice?
```c++
operator bool() noexcept
{
    return static_cast<bool>(m_timer);
}
```

Comment (dmachaj) [reply include/wil/com.h:3403]: I ended up going with option 1 and 3 both.  The nothrow variant can use the bool operator to determine if it is active.  There are throwing and failfast options that will do their thing if construction fails.

**Review by dunhor — APPROVED** [review]

**Comment by dunhor** [comment]
I noticed the `CppWinRTComServerTests::NotifierAndRegistration` failure in other pipeline runs; currently looking at it. This is not a new failure

**Review by dunhor — APPROVED** [review]

File: tests/ComTests.cpp
Comment (brialmsft) [inline tests/ComTests.cpp:3080]: Inproc calls don't go through RPC. COM does use NDR (implemented by rpcrt4.dll) to translate the call into request/response messages. But NDR is generally considered distinct from the RPC runtime. And in the inproc case, we don't use the RPC runtime to transport messages between client and server.

Consider "These test cases require calling a COM server via proxy, so that we can exercise cancellation through the COM runtime."

Normally wouldn't quibble, but comments in public testcases are a bit documentationy.

File: tests/ComTests.cpp
Comment (brialmsft) [inline tests/ComTests.cpp:3085]: Per previous comment, consider:
"To ensure the call goes through a proxy, the object must be non-agile and the call must cross between apartments (MTA -> STA in this case)".

File: tests/ComTests.cpp
Comment (brialmsft) [inline tests/ComTests.cpp:3125]: Would arguably be more straightforward to use agile_ref to create the proxy, rather than activation:
* Add an agile_ref<IStringable> to the MTA scope, ref-captured by the lambda
* On the STA thread, use make_agile to wrap the object in the agile_ref. This replaces CoRegisterClassObject.
* On the MTA thread, use <agileRef>.get() to unwrap the agile_ref. The result is a proxy connected to the STA object. This replaces the CoCreateInstance.

File: tests/ComTests.cpp
Comment (dmachaj) [reply tests/ComTests.cpp]: Copied the suggestion verbatim.

File: tests/ComTests.cpp
Comment (dmachaj) [reply tests/ComTests.cpp]: Copied the suggestion verbatim.

File: tests/ComTests.cpp
Comment (dmachaj) [reply tests/ComTests.cpp]: I was gonna say "that doesn't sound much simpler" but then I tried coding it up and... yeah its simpler.  It also saves me from having that extremely ugly CoCreate polling loop.

**Review by brialmsft — APPROVED** [review]

**Pull request merged**
